File: mobile/android/app/src/main/java/com/dcouple/pane/mobile/SecureStorePlugin.java
package com.dcouple.pane.mobile;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SecureStore")
public class SecureStorePlugin extends Plugin {
  private EncryptedSharedPreferences preferences;
  private EncryptedSharedPreferences preferences() throws Exception {
    if (preferences == null) {
      MasterKey key = new MasterKey.Builder(getContext()).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build();
      preferences = (EncryptedSharedPreferences) EncryptedSharedPreferences.create(getContext(), "pane-secure-store", key, EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV, EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM);
    }
    return preferences;
  }
  @PluginMethod public void get(PluginCall call) { try { String key = call.getString("key"); if (key == null) { call.reject("A key is required"); return; } JSObject result = new JSObject(); String value = preferences().getString(key, null); if (value != null) result.put("value", value); call.resolve(result); } catch (Exception error) { call.reject("Secure storage read failed", error); } }
  @PluginMethod public void set(PluginCall call) { try { String key = call.getString("key"); String value = call.getString("value"); if (key == null || value == null) { call.reject("A key and value are required"); return; } preferences().edit().putString(key, value).apply(); call.resolve(); } catch (Exception error) { call.reject("Secure storage write failed", error); } }
  @PluginMethod public void remove(PluginCall call) { try { String key = call.getString("key"); if (key == null) { call.reject("A key is required"); return; } preferences().edit().remove(key).apply(); call.resolve(); } catch (Exception error) { call.reject("Secure storage delete failed", error); } }
}

File: mobile/ios/App/App/SecureStorePlugin.swift
import Foundation
import Capacitor
import Security

@objc(SecureStorePlugin)
public class SecureStorePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SecureStorePlugin"
    public let jsName = "SecureStore"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise)
    ]
    private let service = "com.dcouple.pane.mobile.secure-store"

    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else { call.reject("A key is required"); return }
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: key, kSecReturnData as String: true]
        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { call.resolve([:]); return }
        guard status == errSecSuccess, let data = result as? Data, let value = String(data: data, encoding: .utf8) else { call.reject("Secure storage read failed"); return }
        call.resolve(["value": value])
    }
    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else { call.reject("A key and value are required"); return }
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: key]
        let attributes: [String: Any] = [kSecValueData as String: Data(value.utf8), kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly]
        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if status == errSecItemNotFound {
            let addStatus = SecItemAdd((query.merging(attributes) { _, new in new }) as CFDictionary, nil)
            guard addStatus == errSecSuccess else { call.reject("Secure storage write failed"); return }
        } else if status != errSecSuccess { call.reject("Secure storage write failed"); return }
        call.resolve()
    }
    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else { call.reject("A key is required"); return }
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: key]
        SecItemDelete(query as CFDictionary); call.resolve()
    }
}
